### ISSUE TITLE
gc3pie crash when running into timeout

### ISSUE THREAD

**Issue by jluethi** [TissueMAPS/TissueMAPS#110]
I'm working on a basically up-to-date TissueMaps instance (v0.3.3, master branch on December 14, 2017) and I'm running into a GC3Pie issue. When I submitted a job that ran into a timeout, the timeout caused error messages in the uwsgi log (see below). Instead of terminating the jobs that ran into a timeout, they are still listed as running and new jobs get submitted (but can't run because the old ones where not killed). I don't know whether anything is actually running or TissueMaps is stuck now.

@riccardomurri do you know where the issue is? It seems related to some IDs that it can't find, see "Could not kill job '23105'. It refers to non-existent local process 4478."

Here is the relevant error log, just after hitting the timeout:
```
Brutally killing worker 1 (pid: 23757)...
Brutally killing worker 2 (pid: 23758)...
SIGINT/SIGQUIT received...killing workers...
Wed Feb  7 18:57:04 2018 - stopping gevent signals watchers for worker 2 (pid: 23758)...
Wed Feb  7 18:57:04 2018 - stopping gevent signals watchers for worker 1 (pid: 23757)...
Wed Feb  7 18:57:04 2018 - stopping gevent sockets watchers for worker 1 (pid: 23757)...
Wed Feb  7 18:57:04 2018 - stopping gevent sockets watchers for worker 2 (pid: 23758)...
Wed Feb  7 18:57:04 2018 - main gevent watchers stopped for worker 1 (pid: 23757)...
Wed Feb  7 18:57:04 2018 - main gevent watchers stopped for worker 2 (pid: 23758)...
Brutally killing worker 4 (pid: 23760)...
Wed Feb  7 18:57:04 2018 - stopping gevent signals watchers for worker 4 (pid: 23760)...
Wed Feb  7 18:57:04 2018 - stopping gevent sockets watchers for worker 4 (pid: 23760)...
Wed Feb  7 18:57:04 2018 - main gevent watchers stopped for worker 4 (pid: 23760)...
goodbye to the gevent Hub on worker 2 (pid: 23758)
goodbye to the gevent Hub on worker 1 (pid: 23757)
goodbye to the gevent Hub on worker 4 (pid: 23760)
[ 23759] 2018-02-07 18:57:04 | ERROR    | gc3.gc3libs                              | Could not kill job '23105'. It refers to non-existent local process 4478. 
[ 23759] 2018-02-07 18:57:04 | ERROR    | gc3.gc3libs                              | Ignoring error in updating state of task '23105': InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.0iLcqi/.gc3pie_shellcmd/resource_usage.txt` for task `23105`: 'shellcmd_user_time'
Traceback (most recent call last):
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 1470, in progress
    self._core.update_job_state(task)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 424, in update_job_state
    **extra_args)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 462, in __update_application
    raise ex
InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.0iLcqi/.gc3pie_shellcmd/resource_usage.txt` for task `23105`: 'shellcmd_user_time'
[ 23759] 2018-02-07 18:57:04 | ERROR    | gc3.gc3libs                              | Ignoring error in updating state of task '23106': InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.Cz8bth/.gc3pie_shellcmd/resource_usage.txt` for task `23106`: 'shellcmd_user_time'
Traceback (most recent call last):
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 1470, in progress
    self._core.update_job_state(task)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 424, in update_job_state
    **extra_args)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 462, in __update_application
    raise ex
InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.Cz8bth/.gc3pie_shellcmd/resource_usage.txt` for task `23106`: 'shellcmd_user_time'
[ 23759] 2018-02-07 18:57:04 | ERROR    | gc3.gc3libs                              | Ignoring error in updating state of task '23107': InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.PoWOlD/.gc3pie_shellcmd/resource_usage.txt` for task `23107`: 'shellcmd_user_time'
Traceback (most recent call last):
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 1470, in progress
    self._core.update_job_state(task)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 424, in update_job_state
    **extra_args)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 462, in __update_application
    raise ex
InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.PoWOlD/.gc3pie_shellcmd/resource_usage.txt` for task `23107`: 'shellcmd_user_time'
[ 23759] 2018-02-07 18:57:04 | ERROR    | gc3.gc3libs                              | Ignoring error in updating state of task '23108': InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.J7CMb9/.gc3pie_shellcmd/resource_usage.txt` for task `23108`: 'shellcmd_user_time'
Traceback (most recent call last):
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 1470, in progress
    self._core.update_job_state(task)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 424, in update_job_state
    **extra_args)
  File "/home/tissuemaps/.local/lib/python2.7/site-packages/gc3libs/core.py", line 462, in __update_application
    raise ex
InvalidValue: Could not open wrapper file `/var/tmp/gc3libs.J7CMb9/.gc3pie_shellcmd/resource_usage.txt` for task `23108`: 'shellcmd_user_time'
worker 4 buried after 1 seconds
Brutally killing worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - stopping gevent signals watchers for worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - stopping gevent sockets watchers for worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - main gevent watchers stopped for worker 3 (pid: 23759)...
Brutally killing worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - stopping gevent signals watchers for worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - stopping gevent sockets watchers for worker 3 (pid: 23759)...
Wed Feb  7 18:57:05 2018 - main gevent watchers stopped for worker 3 (pid: 23759)...
goodbye to the gevent Hub on worker 3 (pid: 23759)
worker 1 buried after 2 seconds
worker 2 buried after 2 seconds
worker 3 buried after 2 seconds
goodbye to uWSGI.
VACUUM: pidfile removed.
VACUUM: unix socket /var/run/uwsgi/tissuemaps.sock removed.
*** Starting uWSGI 2.0.15 (64bit) on [Wed Feb  7 18:57:07 2018] ***
compiled with version: 5.4.0 20160609 on 14 December 2017 15:57:28
os: Linux-4.4.0-96-generic #119-Ubuntu SMP Tue Sep 12 14:59:54 UTC 2017
nodename: joel-big-standalone-001
machine: x86_64
clock source: unix
pcre jit disabled
detected number of CPU cores: 16
current working directory: /
writing pidfile to /var/run/uwsgi/tissuemaps.pid
detected binary path: /home/tissuemaps/.local/bin/uwsgi
your processes number limit is 251598
your memory page size is 4096 bytes
detected max file descriptor number: 1024
- async cores set to 100 - fd table size: 1024
lock engine: pthread robust mutexes
thunder lock: disabled (you can enable it with --thunder-lock)
uwsgi socket 0 bound to TCP address :8080 fd 4
uwsgi socket 1 bound to UNIX address /var/run/uwsgi/tissuemaps.sock fd 5
Python version: 2.7.12 (default, Nov 20 2017, 18:23:56)  [GCC 5.4.0 20160609]
Python main interpreter initialized at 0x17bf5f0
python threads support enabled
your server socket listen backlog is limited to 100 connections
your mercy for graceful operations on workers is 60 seconds
mapped 5464320 bytes (5336 KB) for 400 cores
*** Operational MODE: preforking+async ***
[ 31424] 2018-02-07 18:57:09 | INFO     | tmserver.extensions.gc3pie               | initialize GC3Pie extension
WSGI app 0 (mountpoint='') ready in 4 seconds on interpreter 0x17bf5f0 pid: 31424 (default app)
*** uWSGI is running in multiple interpreter mode ***
spawned uWSGI master process (pid: 31424)
spawned uWSGI worker 1 (pid: 31466, cores: 100)
*** running gevent loop engine [addr:0x493e60] ***
spawned uWSGI worker 2 (pid: 31467, cores: 100)
spawned uWSGI worker 3 (pid: 31468, cores: 100)
spawned uWSGI worker 4 (pid: 31469, cores: 100)
```

**Comment by riccardomurri** [comment]
I think it's the same issue as on Chimp; can you please update to TM "develop" branch and see if you hit the same issue?

**Comment by jluethi** [comment]
With this you mean switch all libraries to the current develop libraries or just some specific ones (e.g. tmlibrary)? And do I need to restart nginx & uwsgi before or after updating the libraries? Does a restart clean up the currently running jobs from the database or do I need to do that manually?

**Comment by riccardomurri** [comment]
Yup, all of them:

    service uwsgi stop
    for code in tmlibrary tmserver jtlibrary; do (cd $code && git fetch && git checkout develop); done
   service uwsgi start

You might also need to upgrade GC3Pie to master:

    pip install --user 'https://github.com/uzh/gc3pie.git#egg=gc3pie'

**Comment by jluethi** [comment]
Ok, thanks for the instructions, I've done that. I still see the old jobs as running & submitted. Can I  somehow force TissueMaps to make all currently running jobs fail? Or do I have to remove them manually from the database somehow?

**Comment by riccardomurri** [comment]
Please do not remove rows from the DB -- that could wreak havoc on GC3Pie's idea of how jobs run, and make debugging impossible.

To kill *all* jobs, run the following command (need to have GC3Pie "master" installed)::

    gkill -s 'postgresql://tissuemaps@localhost/tissuemaps#table=tasks' -A

If you want to kill jobs selectively, run this command to map out the jobs currently in the database::

    gsession list -r 'postgresql://tissuemaps@localhost/tissuemaps#table=tasks'

To inspect a particular job, please run::

    ginfo -v -s 'postgresql://tissuemaps@localhost/tissuemaps#table=tasks' $job_id

**Comment by jluethi** [comment]
Thanks @riccardomurri 
This has allowed me to stop all my jobs and thus work around the problem with the timeouts.

**Comment by jluethi** [comment]
@riccardomurri On the current develop branch, running 
`gkill -s 'postgresql://tissuemaps@localhost/tissuemaps#table=tasks' -A`
gives the following error message:
`gkill: [2018-05-29 11:35:18] CRITICAL: RuntimeError: Session postgresql://tissuemaps@localhost/tissuemaps#table=tasks not found`

I have a timeout issue again and would need to kill the jobs. How do I do this now? What changed?
(Running this on my tissuemaps server while logged in as tissuemaps user, who has access to the db (at least the read access is working fine).

**Comment by riccardomurri** [comment]
This should have been fixed in the array of commits that led to v0.4.0
